crates/uv/src/commands/pip/tree.rs

```diff
@@ -0,0 +1,210 @@
+    // Set of package names that are required by at least one installed distribution.
+    // It is used to determine the starting nodes when recursing the
+    // dependency graph.
+    required_packages: HashSet<PackageName>,
+}
+
+impl<'a> DisplayDependencyGraph<'a> {
+    /// Create a new [`DisplayDependencyGraph`] for the set of installed distributions.
+    fn new(site_packages: &'a SitePackages) -> DisplayDependencyGraph<'a> {
+        let mut dist_by_package_name = HashMap::new();
+        let mut required_packages = HashSet::new();
+        for site_package in site_packages.iter() {
+            dist_by_package_name.insert(site_package.name(), site_package);
+        }
+        for site_package in site_packages.iter() {
+            for required in required_with_no_extra(site_package) {
+                required_packages.insert(required.name.clone());
+            }
+        }
+
+        Self {
+            site_packages,
+            dist_by_package_name,
+            required_packages,
+        }
+    }
+
+    // Depth-first traversal of the given distribution and its dependencies.
+    fn visit(
+        &self,
+        installed_dist: &InstalledDist,
+        visited: &mut HashSet<String>,
+        path: &mut Vec<String>,
+    ) -> Result<Vec<String>, Error> {
+        let mut lines = Vec::new();
+
+        let package_name = installed_dist.name().to_string();
+        if path.contains(&package_name) {
+            let cycle_start_index = path.iter().position(|r| r == &package_name).unwrap();
+            return Err(Error::CyclicDependency(path[cycle_start_index..].to_vec()));
+        }
+        let is_visited = visited.contains(&package_name);
+        lines.push(render_line(installed_dist, path.len(), is_visited));
+        if is_visited {
+            return Ok(lines);
+        }
+
+        path.push(package_name.clone());
+        visited.insert(package_name.clone());
+        for required in &required_with_no_extra(installed_dist) {
+            if self.dist_by_package_name.contains_key(&required.name) {
+                match self.visit(self.dist_by_package_name[&required.name], visited, path) {
+                    Ok(visited_lines) => lines.extend(visited_lines),
+                    Err(e) => {
+                        return Err(e);
+                    }
+                }
+            }
+        }
+        path.pop();
+        Ok(lines)
+    }
+
+    // Depth-first traverse the nodes to render the tree.
+    // The starting nodes are the ones without incoming edges.
+    fn render(&self) -> Result<Vec<String>, Error> {
+        let mut visited: HashSet<String> = HashSet::new();
+        let mut path: Vec<String> = Vec::new();
+        let mut lines: Vec<String> = Vec::new();
+        for site_package in self.site_packages.iter() {
+            if !self.required_packages.contains(site_package.name()) {
+                match self.visit(site_package, &mut visited, &mut path) {
```

Comment:
Should we be reusing the same `path` for all dependencies? Cycles can only occur in `visit`, so it seems like we should be clearing the `Vec` or creating a new `path` for each dependency.